packages/runtime/src/runtimes/react/controls/link.ts

```diff
@@ -0,0 +1,102 @@
+import { MouseEvent } from 'react'
+
+import scrollIntoView from 'scroll-into-view-if-needed'
+import { useElementId } from '../'
+import { usePage } from '../../../components/hooks'
+import { LinkControlData, LinkControlDefinition } from '../../../controls'
+
+type LinkValue = {
+  href: string
+  target: '_blank' | '_self' | undefined
+  onClick: (event: MouseEvent<HTMLElement>) => void
+}
+
+export type LinkControlValue<T extends LinkControlDefinition> =
+  undefined extends T['config']['defaultValue'] ? LinkValue | undefined : LinkValue
+
+export function useLinkControlValue<T extends LinkControlDefinition>(
+  data: LinkControlData | undefined,
+  definition: T,
+): LinkControlValue<T> {
+  const link = data ?? definition.config.defaultValue
+  const page = usePage(link && link.type === 'OPEN_PAGE' ? link.payload.pageId : null)
+  const elementKey =
+    link?.type === 'SCROLL_TO_ELEMENT' ? link.payload.elementIdConfig?.elementKey : null
+  const elementId = useElementId(elementKey)
+
+  let href = '#'
+  let target: '_blank' | '_self' | undefined
+  let block: 'start' | 'center' | 'end' | undefined
+
+  if (link) {
+    switch (link.type) {
+      case 'OPEN_PAGE': {
+        if (page) href = `/${page.pathname}`
+
+        target = link.payload.openInNewTab ? '_blank' : '_self'
+
+        break
+      }
+
+      case 'OPEN_URL': {
+        href = link.payload.url
+
+        target = link.payload.openInNewTab ? '_blank' : '_self'
+
+        break
+      }
+
+      case 'SEND_EMAIL': {
+        const { to, subject = '', body = '' } = link.payload
+
+        if (to != null) href = `mailto:${to}?subject=${subject}&body=${body}`
+
+        break
+      }
+
+      case 'CALL_PHONE': {
+        href = `tel:${link.payload.phoneNumber}`
+
+        break
+      }
+
+      case 'SCROLL_TO_ELEMENT': {
+        href = `#${elementId ?? ''}`
+        block = link.payload.block
+
+        break
+      }
+
+      default:
+        throw new RangeError(`Invalid link type "${(link as any).type}."`)
+    }
+  }
+
+  function handleClick(event: MouseEvent<HTMLElement>) {
+    if (event.defaultPrevented) return
+
+    if (link && link.type === 'SCROLL_TO_ELEMENT') {
+      let hash: string | undefined
+
+      try {
+        hash = new URL(`http://www.example.com/${href}`).hash
+      } catch (error) {
+        console.error(`Link received invalid href: ${href}`, error)
+      }
+
+      if (href != null && href === hash) {
+        event.preventDefault()
+        const view = event.view as unknown as Window
+
+        scrollIntoView(view.document.querySelector(hash)!, {
+          behavior: 'smooth',
+          block,
+        })
+
+        if (view.location.hash !== hash) view.history.pushState({}, '', hash)
+      }
+    }
+  }
+
+  return { href, target, onClick: handleClick }
```

Comment:
What do you think about memoizing this so that if `href` and `target` don't change, the click handler and the returned object also don't change?

Reply:
Good idea 👍